packages/backend-tasks/src/tasks/types.ts

```diff
@@ -0,0 +1,138 @@
+ */
+export type TaskFunction =
+  | ((abortSignal: AbortSignal) => void | Promise<void>)
+  | (() => void | Promise<void>);
+
+/**
+ * Options that apply to the invocation of a given task.
+ *
+ * @public
+ */
+export interface TaskDefinition {
+  /**
+   * A unique ID (within the scope of the plugin) for the task.
+   */
+  id: string;
+
+  /**
+   * The actual task function to be invoked regularly.
+   */
+  fn: TaskFunction;
+
+  /**
+   * An abort signal that, when triggered, will stop the recurring execution of
+   * the task.
+   */
+  signal?: AbortSignal;
+
+  /**
+   * The maximum amount of time that a single task invocation can take, before
+   * it's considered timed out and gets "released" such that a new invocation
+   * is permitted to take place (possibly, then, on a different worker).
+   *
+   * If no value is given for this field then there is no timeout. This is
+   * potentially dangerous.
+   */
+  timeout: Duration;
+
+  /**
+   * The amount of time that should pass between task invocation starts.
+   * Essentially, this equals roughly how often you want the task to run.
+   *
+   * This is a best effort value; under some circumstances there can be
+   * deviations. For example, if the task runtime is longer than the frequency
+   * and the timeout has not been given or not been exceeded yet, the next
+   * invocation of this task will be delayed until after the previous one
+   * finishes.
+   *
+   * The system does its best to avoid overlapping invocations.
+   *
+   * If no value is given for this field then the task will only be invoked
+   * once (on any worker) and then unscheduled automatically.
+   */
+  frequency: Duration;
+
+  /**
+   * The amount of time that should pass before the first invocation happens.
+   *
+   * This can be useful in cold start scenarios to stagger or delay some heavy
+   * compute jobs.
+   *
+   * If no value is given for this field then the first invocation will happen
+   * as soon as possible.
+   */
+  initialDelay?: Duration;
+}
+
+/**
+ * Deals with the scheduling of distributed tasks, for a given plugin.
+ *
+ * @public
+ */
+export interface PluginTaskScheduler {
+  /**
+   * Schedules a task function for coordinated exclusive invocation across
+   * workers.
+   *
+   * If the task was already scheduled since before by us or by another party,
+   * its options are just overwritten with the given options, and things
+   * continue from there.
+   *
+   * @param definition - The task definition
+   */
+  scheduleTask(task: TaskDefinition): Promise<void>;
+}
+
+function isValidOptionalDurationString(d: string | undefined): boolean {
+  try {
+    return !d || Duration.fromISO(d).isValid === true;
+  } catch {
+    return false;
+  }
+}
+
+export const taskSettingsV1Schema = z.object({
+  version: z.literal(1),
+  initialDelayDuration: z
+    .string()
+    .optional()
+    .refine(isValidOptionalDurationString, { message: 'Invalid duration' }),
+  recurringAtMostEveryDuration: z
```

Comment:
Would it make sense to simply call this `frequency` or similar so we don't need to necessarily make assumptions how it is used to calculate the next recurrence/run. Thinking that a cron string can be used here also and then used in `releaseTask` to calculate the next run.

Reply:
Hm, I see where you're going. The intent here was to actually be as explicit as possible, since we might want to add more types of recurrency side by side with this, going forward. Kinda like [java's scheduled executor](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html) distinguishes between `scheduleAtFixedRate` and `scheduleWithFixedDelay`, for example, which we may want to mimic later on. Didn't want to give the impression that this was a strict promise, that the duration is always held true whether the task finishes in time or not.